// soil-ts/_internal/_global.jsx
export var pathAppData = Folder.appData;
export var pathAppPackage = Folder.appPackage;
export var pathCommonFiles = Folder.commonFiles;
export var pathCurrent = Folder.current;
export var pathDesktop = Folder.desktop;
export var pathFs = Folder.fs;
export var pathMyDocuments = Folder.myDocuments;
export var pathScript = $.fileName;
export var pathStartup = Folder.startup;
export var pathSystem = Folder.system;
export var pathTemp = Folder.temp;
export var pathTrash = Folder.trash;
export var pathUserData = Folder.userData;
export var IS_KEY_LABEL_EXISTS = parseFloat(app.version) > 22.5;
export var PROPERTY_INTERPOLATION_TYPE = [6612, 6613, 6614];
export var reScriptFileName = /\.(js|jsx|jsxbin)$/i;
export var rePrefsKeyName = /"([^"]+?)"/;
export var reSplitPrefsBlock = /^\s*?"([^"]*?)"\s*?=\s*?\s*/gm;
export var reSplitPrefsSection = /^\["([^\]]+)"\][^"]/gm;
export var reTemplateString = /\$\{(\d+)\}/g;
export var jsonEscapes = { "\b": "\\b", "\t": "\\t", "\n": "\\n", "\f": "\\f", "\r": "\\r", "\v": "\\v", '"': '\\"', "\\": "\\\\" };
export var reEscapedJson = /[\\"\u0000-\u001f\u007f-\u009f\u00ad\u0600-\u0604\u070f\u17b4\u17b5\u200c-\u200f\u2028-\u202f\u2060-\u206f\ufeff\ufff0-\uffff]/g;
export var reHasEscapedJson = new RegExp(reEscapedJson.source);
